duplex_http_call.go

```diff
@@ -316,3 +313,81 @@ func cloneURL(oldURL *url.URL) *url.URL {
 	}
 	return newURL
 }
+
+// payloadCloser is a ReadCloser that wraps a bytes.Buffer. It's used to
+// implement the request body for unary calls. To safely reuse the buffer
+// call Wait after the response is received to ensure the payload has been
+// fully read.
+type payloadCloser struct {
+	wait sync.WaitGroup
+
+	mu     sync.Mutex
+	buf    *bytes.Buffer
+	readN  int
+	isDone bool // true if the payload has been fully read
+}
+
+func newPayloadCloser(buf *bytes.Buffer) *payloadCloser {
+	payload := &payloadCloser{
+		buf: buf,
+	}
+	payload.wait.Add(1) // Wait until complete is called
+	return payload
+}
+
+// Read implements io.Reader.
+func (p *payloadCloser) Read(dst []byte) (readN int, err error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.buf == nil {
+		return 0, io.EOF
+	}
+	src := p.buf.Bytes()
+	readN = copy(dst, src[p.readN:])
+	p.readN += readN
+	if readN == 0 && p.readN == len(src) {
+		p.completeWithLock()
+		err = io.EOF
+	}
+	return readN, err
+}
+
+// Close implements io.Closer. It signals that the payload has been fully read.
+func (p *payloadCloser) Close() error {
```

Comment:
Close should be called by the `http.Transport` on receiving an `io.EOF` and finished with the request body. But under testing  the HTTP2 transport would sometimes delay the call leading to a deadlock on `Wait()`. We therefore also signal completion by reading all the bytes in the buffer.